aleph/logic/api_keys.py

```diff
@@ -0,0 +1,108 @@
+import datetime
+
+from flask import render_template
+from sqlalchemy import and_, or_, func
+
+from aleph.core import db
+from aleph.model import Role
+from aleph.model.common import make_token
+from aleph.logic.mail import email_role
+from aleph.logic.roles import update_role
+from aleph.logic.util import ui_url
+
+# Number of days after which API keys expire
+API_KEY_EXPIRATION_DAYS = 90
+
+# Number of days before an API key expires
+API_KEY_EXPIRES_SOON_DAYS = 7
+
+
+def generate_user_api_key(role):
+    event = "regenerated" if role.has_api_key else "generated"
+    params = {"role": role, "event": event}
+    plain = render_template("email/api_key_generated.txt", **params)
+    html = render_template("email/api_key_generated.html", **params)
+    subject = f"API key {event}"
+    email_role(role, subject, html=html, plain=plain)
+
+    now = datetime.datetime.utcnow().replace(microsecond=0)
+    role.api_key = make_token()
+    role.api_key_expires_at = now + datetime.timedelta(days=API_KEY_EXPIRATION_DAYS)
+    role.api_key_expiration_notification_sent = None
+
+    db.session.add(role)
+    db.session.commit()
+    update_role(role)
+
+    return role.api_key
+
+
+def send_api_key_expiration_notifications():
```

Comment:
This method sends notifications to users with an API key that will expire soon (in less than 7 days) or that has expired recently.

But you were probably asking because this function isn’t called anywhere right now and indeed I did forget to push one additional change that’s required to execute this daily.

Reply:
Done